README.md
# Mapeo View Components

These components are designed for viewing data in Mapeo. They share a common interface:

### Common Props

| Name                | Type                                                                                           | Default | Description                                                                                                                                                                                                                   |
| ------------------- | ---------------------------------------------------------------------------------------------- | ------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| observations        | [`Observation[]`](https://github.com/digidem/mapeo-schema/blob/master/schema/observation.json) | `[]`    | Array of Mapeo observations                                                                                                                                                                                                   |
| onUpdateObservation | `func`                                                                                         |         | Callback fired when an observation has been updatedby the view. <br /> <br /> **Signature:** <br /> `(observation: Observation) => void` _`observation:`_ The updated observation                                             |
| presets             | [`Preset[]`](https://github.com/digidem/mapeo-schema/blob/master/schema/preset.json)           | `[]`    | Array of Mapeo [`Preset`](https://github.com/digidem/mapeo-schema/blob/master/schema/preset.json)s with an array of [`Field`](https://github.com/digidem/mapeo-schema/blob/master/schema/field.json)s instead of Field `id`s. |
| filter              | `array`                                                                                        |         | [Filter expression](https://github.com/digidem/mapeo-entity-filter#filter-expression-language) used to filter the observations that will be shown.                                                                            |
| getMediaUrl         | `func`                                                                                         |         | Function called with an `id` of an image attachment and a `size`, should return a valid URL to the image. <br /> <br /> **Signature:** <br /> `(id: string, size: 'thumbnail' | 'preview' | 'original') => string`            |
| getIconUrl          | `func`                                                                                         |         | Function called with an `id` of an icon, should return a valid URL to the icon.<br /> <br /> **Signature:** <br /> `(id: string) => string`                                                                                   |

## Contents

- [`<MapView />`](#mapview-)
  - [MapView Props](#mapview-props)
- [`<MediaView />`](#mediaview-)
  - [MediaView Props](#mediaview-props)
- [`<ReportView />`](#reportview-)
  - [ReportView Props](#reportview-props)

## `<MapView />`

Displays observations on a map.

### MapView Props

| Name               | Type     | Default                                 | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| ------------------ | -------- | --------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| onMapMove          | `func`   |                                         | Called with [CameraOptions](https://docs.mapbox.com/mapbox-gl-js/api/#cameraoptions) with properties `center`, `zoom`, `bearing`, `pitch` whenever the map is moved                                                                                                                                                                                                                                                                                                  |
| initialMapPosition | `object` |                                         | Initial [`CameraOptions`](https://docs.mapbox.com/mapbox-gl-js/api/#cameraoptions) position for map - an object with properties `center`, `zoom`, `bearing`, `pitch`. If this is not set then the map will by default zoom to the bounds of the observations. If you are going to unmount and re-mount this component (e.g. within tabs) then you will want to use onMove to store the position in state, and pass it as initialPosition for when the map re-mounts. |
| mapStyle           | `string` | `'mapbox://styles/mapbox/outdoors-v10'` | A [Mapbox Style URL](https://docs.mapbox.com/help/glossary/style-url/)                                                                                                                                                                                                                                                                                                                                                                                               |
| mapboxAccessToken  | `string` |                                         | A [Mapbox Access Token](https://docs.mapbox.com/help/glossary/access-token/) used to access the style                                                                                                                                                                                                                                                                                                                                                                |

### MapView instance methods

#### `flyTo({center, zoom}, eventData?)`

Changes any combination of center, zoom, bearing, and pitch, animating the transition along a curve that evokes flight. The animation seamlessly incorporates zooming and panning to help the user maintain her bearings even after traversing a great distance, takes the same options as the
[`flyTo` method of
`mapbox-gl-js`](https://docs.mapbox.com/mapbox-gl-js/api/#map#flyto)

#### `fitBounds(bounds, options?, eventData?)`

Pans and zooms the map to contain its visible area within the specified geographical bounds. This function will also reset the map's bearing to 0 if bearing is nonzero, takes the same options as the
[`fitBounds` method of
`mapbox-gl-js`](https://docs.mapbox.com/mapbox-gl-js/api/#map#fitbounds)

## `<MediaView />`

Display a grid of all the media attachments from the observations.

### MediaView Props

MediaView does not currently have any additional props beyond the common props
above.

## `<ReportView />`

Display observations as a report that can be printed.

### ReportView Props

ReportView shares several props with [MapView](#mapview-). These props apply to
the inset map in the ReportView.

| Name               | Type     | Default                                 | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| ------------------ | -------- | --------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| onMapMove          | `func`   |                                         | Called with [CameraOptions](https://docs.mapbox.com/mapbox-gl-js/api/#cameraoptions) with properties `center`, `zoom`, `bearing`, `pitch` whenever the map is moved                                                                                                                                                                                                                                                                                                  |
| initialMapPosition | `object` |                                         | Initial [`CameraOptions`](https://docs.mapbox.com/mapbox-gl-js/api/#cameraoptions) position for map - an object with properties `center`, `zoom`, `bearing`, `pitch`. If this is not set then the map will by default zoom to the bounds of the observations. If you are going to unmount and re-mount this component (e.g. within tabs) then you will want to use onMove to store the position in state, and pass it as initialPosition for when the map re-mounts. |
| mapStyle           | `string` | `'mapbox://styles/mapbox/outdoors-v10'` | A [Mapbox Style URL](https://docs.mapbox.com/help/glossary/style-url/)                                                                                                                                                                                                                                                                                                                                                                                               |
| mapboxAccessToken  | `string` |                                         | A [Mapbox Access Token](https://docs.mapbox.com/help/glossary/access-token/) used to access the style                                                                                                                                                                                                                                                                                                                                                                |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.2.5](https://github.com/digidem/react-mapfilter/compare/v3.2.3...v3.2.5) (2020-07-31)


### Bug Fixes

* Editing multiselect fields with single values shouldn't cause ([85c4dbc](https://github.com/digidem/react-mapfilter/commit/85c4dbcfd6c29e88cace5644874f7fcd4e81b1ee))
* Fix flow types not being correctly removed from compiled code ([cc2cfb5](https://github.com/digidem/react-mapfilter/commit/cc2cfb597e625dd922ca8dd5ef0e7ac5d176a480))

### [3.2.4](https://github.com/digidem/react-mapfilter/compare/v3.2.3...v3.2.4) (2020-07-16)

### Bug Fixes

- Fix flow types not being correctly removed from compiled code ([cc2cfb5](https://github.com/digidem/react-mapfilter/commit/cc2cfb597e625dd922ca8dd5ef0e7ac5d176a480))

### [3.2.3](https://github.com/digidem/react-mapfilter/compare/v3.2.2...v3.2.3) (2020-07-14)

### Bug Fixes

- fix failing test for converting btween non-string values and labels ([79561db](https://github.com/digidem/react-mapfilter/commit/79561db2dc220e47d33ef2e6108720a51d93bdff))
- Fix number fields to always return number not string ([01968ca](https://github.com/digidem/react-mapfilter/commit/01968caffa58d9623d40a7a472a80e1a4044ea4a))
- Select should properly respect value/label objects for options ([60aeb28](https://github.com/digidem/react-mapfilter/commit/60aeb28ed73ab32653d8a3a60857d5ece2ad9af7))

### [3.2.2](https://github.com/digidem/react-mapfilter/compare/v3.2.1...v3.2.2) (2020-07-09)

### Bug Fixes

- Ensure TextField fallback has handleChange property ([801dfcd](https://github.com/digidem/react-mapfilter/commit/801dfcd6fc3bec3e5efa5919fac425695de3c2f1))
- SelectMultiple should not throw errors in console ([f5a219d](https://github.com/digidem/react-mapfilter/commit/f5a219d4ca7a74fb7608928f7564e9d8e982c1e7))

### [3.2.1](https://github.com/digidem/react-mapfilter/compare/v3.2.0...v3.2.1) (2020-07-09)

### Bug Fixes

- Don't automatically expand filters by default ([162722b](https://github.com/digidem/react-mapfilter/commit/162722bf2d77255d7a5eba6455d6e968dc4d3a9e))

## [3.2.0](https://github.com/digidem/react-mapfilter/compare/v3.1.0...v3.2.0) (2020-07-09)

### Features

- use @material-ui/lab/Autocomplete for SelectMultiple ([aced10d](https://github.com/digidem/react-mapfilter/commit/aced10da2724aaa5ea0d04122ccd7d8343fc3148))

## [3.1.0](https://github.com/digidem/react-mapfilter/compare/v3.0.0...v3.1.0) (2020-07-09)

### Features

- Support `localized` type fields ([9dba6f1](https://github.com/digidem/react-mapfilter/commit/9dba6f1d6b9a219d8ccaee04e348e6e615b61aca))
- Support `singleline` text fields ([67295b1](https://github.com/digidem/react-mapfilter/commit/67295b1422a0485de5072a274c8238d2f167df02))

## [3.0.0](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.15...v3.0.0) (2019-11-25)

### Bug Fixes

- Change label from "created" to "date of observation" ([607bcd0](https://github.com/digidem/react-mapfilter/commit/607bcd00bc1353e6d11a6989790f1c121879f91e))

## [3.0.0-beta.15](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.14...v3.0.0-beta.15) (2019-11-23)

### Bug Fixes

- Fix filter bug (can't return to show all) when filtering by two or more fields ([69a8aa3](https://github.com/digidem/react-mapfilter/commit/69a8aa3fa8bf206c852d7e740da7eb66c40b16c1))

## [3.0.0-beta.14](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.13...v3.0.0-beta.14) (2019-11-18)

### Bug Fixes

- Translate Delete Observation menu item ([9faeec9](https://github.com/digidem/react-mapfilter/commit/9faeec96b682b7028c68523cd46c0bccd993d461))

## [3.0.0-beta.13](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.12...v3.0.0-beta.13) (2019-11-12)

### Features

- Add read-only select_multiple field support ([0fc88e3](https://github.com/digidem/react-mapfilter/commit/0fc88e39fbd3fef1b78e62a7c367aeea18e8726c))

## [3.0.0-beta.12](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.11...v3.0.0-beta.12) (2019-10-31)

### Features

- Add delete observation action ([9688317](https://github.com/digidem/react-mapfilter/commit/9688317a1e415a70e73a2d453d4a41a5de0f9aff))

### Bug Fixes

- Remove unused IntlProvider in MapView ([fa5c45b](https://github.com/digidem/react-mapfilter/commit/fa5c45bc145ddb868ef8dc26dac98076bd0f1c67))

## [3.0.0-beta.11](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.10...v3.0.0-beta.11) (2019-10-08)

### Bug Fixes

- Hide fields button ([9659a36](https://github.com/digidem/react-mapfilter/commit/9659a36))

## [3.0.0-beta.10](https://github.com/digidem/react-mapfilter/compare/v3.0.0-beta.9...v3.0.0-beta.10) (2019-10-08)

### Bug Fixes

- broken image color ([85005b3](https://github.com/digidem/react-mapfilter/commit/85005b3))
- don't move map if initialMapPosition is set ([526a5ef](https://github.com/digidem/react-mapfilter/commit/526a5ef))
- Don't show option to hide categoryId and notes from reports ([79d1fa4](https://github.com/digidem/react-mapfilter/commit/79d1fa4))
- Fix date filter not including max day ([dd52013](https://github.com/digidem/react-mapfilter/commit/dd52013))
- fix hiding fields in Report view ([24e3d44](https://github.com/digidem/react-mapfilter/commit/24e3d44))
- Fix report toolbar positioning ([0516d07](https://github.com/digidem/react-mapfilter/commit/0516d07))
- forward map ref ([005a454](https://github.com/digidem/react-mapfilter/commit/005a454))
- Min/max date always less than / greater than other ([1cc7a27](https://github.com/digidem/react-mapfilter/commit/1cc7a27))
- sort media (most recent first) ([3f999c1](https://github.com/digidem/react-mapfilter/commit/3f999c1))
- Sort presets by sort field, then by name ([d64312d](https://github.com/digidem/react-mapfilter/commit/d64312d))

src/types.js
// @flow
import * as React from 'react'
import * as valueTypes from './constants/value_types'
import type { Observation, Preset } from 'mapeo-schema'
// import type { Properties as CSSProperties } from 'csstype'

import type {
  // FeatureCollectionTemplate,
  Bbox,
  Point2D,
  Point3D
} from 'flow-geojson'

/**
 * Observation Attachment
 */
export type Attachment = $ElementType<
  $NonMaybeType<$ElementType<Observation, 'attachments'>>,
  number
>

// export type StyleProp = CSSProperties<string | number>

// Almost JSON value, but we can also have 'undefined' as a value
export type JSONValue =
  | void
  | null
  | number
  | string
  | boolean
  | JSONObject // eslint-disable-line no-use-before-define
  | JSONArray // eslint-disable-line no-use-before-define
export type JSONObject = { [key: string]: JSONValue }
type JSONArray = Array<JSONValue>
type Properties = JSONObject | null

type FlattenedProperties = {
  [key: string]: null | number | string | boolean
} | null
// type FlattenedPropertiesArray = {
//   [key: string]: null | number | string | boolean | JSONArray
// } | null

type FeatureTemplate<G, P> = {
  id?: string | number,
  type: 'Feature',
  bbox?: Bbox,
  properties: P,
  geometry: G
}

type FeatureTemplateWithId<G, Properties> = FeatureTemplate<G, Properties> & {
  id: string | number
}

export type PointFeature = FeatureTemplate<Point2D | Point3D | null, Properties>

export type PointFeatureWithId = FeatureTemplateWithId<
  Point2D | Point3D | null,
  Properties
>

export type FlattenedPointFeature = FeatureTemplate<
  Point2D | Point3D | null,
  FlattenedProperties
>

// export type FeatureCollection = FeatureCollectionTemplate<PointFeature>

export type PaperSize = 'a4' | 'letter'

// Used to store state about field visibility in views
export type FieldState = Array<{|
  id: string,
  hidden: boolean,
  label: React.Node
|}>
export type Filter = Array<number | string | boolean | null | Array<Filter>>

export type ValueTypes = { [fieldkey: string]: $Values<typeof valueTypes> }

export type FieldOrder = { [fieldkey: string]: number }

export type Classes<S> = { [className: $Keys<S>]: string }

export type StringStatistic = {|
  count: number,
  lengthMin?: number,
  lengthMax?: number,
  lengthVariance?: number,
  lengthMean?: number,
  wordsMin?: number,
  wordsMax?: number,
  wordsVariance?: number,
  wordsMean?: number,
  values: Map<string, number>
|}

export type NumberStatistic = {|
  count: number,
  min?: number,
  max?: number,
  variance?: number,
  mean?: number,
  values: Map<number, number>
|}

export type DateStatistic = {|
  count: number,
  min?: string,
  max?: string,
  mean?: string,
  values: Map<string, number>
|}

export type NonArrayFieldStatistic = {|
  string: StringStatistic,
  boolean: {|
    count: number,
    values: Map<boolean, number>
  |},
  number: NumberStatistic,
  date: DateStatistic,
  datetime: DateStatistic,
  url: number,
  image: number,
  video: number,
  audio: number,
  null: number,
  undefined: number,
  location: number
|}

export type FieldStatistic = {|
  ...$Exact<NonArrayFieldStatistic>,
  array?: {|
    count: number,
    lengthMin: number,
    lengthMax: number,
    valueStats: FieldStatistic
  |}
|}

export type Statistics = { [fieldname: string]: FieldStatistic }

type MediaType =
  | typeof valueTypes.IMAGE_URL
  | typeof valueTypes.AUDIO_URL
  | typeof valueTypes.VIDEO_URL
export type MediaArray = Array<{ src: string, type: MediaType }>

export type Coordinates = {
  altitude?: number,
  heading?: number,
  longitude: number,
  speed?: number,
  latitude: number,
  accuracy?: number
}

export type Key = string | Array<string | number>

type BaseField = {|
  // A unique id used to reference the field from presets
  id: string,
  // They key in a tags object that this field applies to. For nested
  // properties, key can be an array e.g. for tags = { foo: { bar: 1 } } the key
  // is ['foo', 'bar']
  key: Key,
  label?: string,
  // Displayed as a placeholder or hint for the field: use for additional
  // context or example responses for the user
  placeholder?: string,
  // If a field definition contains the property "universal": true, this field will appear in the "Add Field" list for all presets
  universal?: boolean,
  // Displayed, but cannot be edited
  readonly?: boolean
|}

// type FieldType =
//   | 'text'
//   | 'number'
//   | 'select_one'
//   | 'select_multiple'
//   | 'date'
//   | 'datetime'

export type TextField = {
  ...BaseField,
  type: 'text' | 'textarea' | 'localized',
  appearance?: 'singleline' | 'multiline',
  // Spaces are replaced with underscores
  snake_case?: boolean
}

export type LinkField = {
  ...BaseField,
  type: 'link'
}

export type NumberField = {
  ...BaseField,
  type: 'number',
  min_value?: number,
  max_value?: number
}

export type SelectableFieldValue = number | string | boolean | null

export type SelectOptions = Array<
  SelectableFieldValue | {| value: SelectableFieldValue, label: string |}
>

export type SelectOneField = {
  ...BaseField,
  type: 'select_one',
  options: SelectOptions,
  // User can enter their own reponse if not on the list (defaults to true on
  // desktop, false on mobile)
  other?: boolean,
  // Spaces are replaced with underscores
  snake_case?: boolean
}

export type SelectMultipleField = {
  ...$Exact<SelectOneField>,
  type: 'select_multiple'
}

export type DateField = {
  ...BaseField,
  type: 'date',
  min_value?: string,
  max_value?: string
}

export type DateTimeField = {
  ...BaseField,
  type: 'datetime',
  min_value?: string,
  max_value?: string
}

export type Field =
  | TextField
  | NumberField
  | SelectOneField
  | SelectMultipleField
  | DateField
  | DateTimeField
  | LinkField

export type MessageDescriptor = {
  id: string,
  description?: string | {},
  defaultMessage?: string
}

type IntlConfig = {|
  locale?: string,
  timeZone?: string,
  textComponent?: any,
  messages?: { [key: string]: string },
  defaultLocale: string,
  onError?: (err: string) => void
|}

type FormatDateOptions = {|
  localeMatcher?: 'lookup' | 'best fit',
  timeZone?: string,
  hour12?: boolean,
  hourCycle?: 'h11' | 'h12' | 'h23' | 'h24',
  formatMatcher?: 'basic' | 'best fit',
  weekday?: 'long' | 'short' | 'narrow',
  era?: 'long' | 'short' | 'narrow',
  year?: 'numeric' | '2-digit',
  month?: 'numeric' | '2-digit' | 'long' | 'short' | 'narrow',
  day?: 'numeric' | '2-digit',
  hour?: 'numeric' | '2-digit',
  minute?: 'numeric' | '2-digit',
  second?: 'numeric' | '2-digit',
  timeZoneName?: 'long' | 'short'
|}

type oneToTwenty =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20

type FormatNumberOptions = {|
  localeMatcher?: 'lookup' | 'best fit',
  style?: 'decimal' | 'currency' | 'percent',
  currenty?: string,
  currencyDisplay?: 'symbol' | 'code' | 'name',
  useGrouping?: boolean,
  minimumIntegerDigits?: oneToTwenty | 21,
  minimumFractionDigits?: 0 | oneToTwenty,
  maximumFractionDigits?: 0 | oneToTwenty,
  minimumSignificantDigits?: oneToTwenty | 21,
  maximumSignificantDigits?: oneToTwenty | 21
|}

type FormatRelativeOptions = {|
  localeMatcher?: 'lookup' | 'best fit',
  numeric?: 'always' | 'auto',
  type?: 'long' | 'short' | 'narrow'
|}

type Unit =
  | 'second'
  | 'minute'
  | 'hour'
  | 'day'
  | 'week'
  | 'month'
  | 'quarter'
  | 'year'

type FormatPluralOptions = {
  type?: 'cardinal' | 'ordinal'
}

export type Primitive = string | boolean | null | void | number

type MessageValues = { [key: string]: Primitive }
type IntlFormatters = {|
  formatDate: (value: number | Date, opts?: FormatDateOptions) => string,
  formatTime: (value: number | Date, opts?: FormatDateOptions) => string,
  formatRelativeTime: (
    value: number,
    unit: Unit,
    opts?: FormatRelativeOptions
  ) => string,
  formatNumber: (value: number, opts?: FormatNumberOptions) => string,
  formatPlural: (value: number, opts?: FormatPluralOptions) => string,
  formatMessage: (
    descriptor: MessageDescriptor,
    values?: MessageValues
  ) => string,
  formatHTMLMessage: (
    descriptor: MessageDescriptor,
    values?: MessageValues
  ) => string
|}

export type IntlShape = {|
  ...IntlConfig,
  ...IntlFormatters
|}

/** A function that receives an observation attachment and should return a URL
 * to retrieve the attachment */
export type GetMedia = (
  attachment: Attachment,
  options?: { width: number, height: number }
) => { src: string, type: 'image' | 'video' | 'audio' } | void

export type GetMediaUrl = (
  attachmentId: string,
  size: 'thumbnail' | 'preview' | 'original'
) => string

export type GetIconUrl = (iconId: string) => string

export type PresetWithFields = {
  ...$Exact<Preset>,
  fields: Field[]
}

export type PresetWithAdditionalFields = {
  ...$Exact<PresetWithFields>,
  additionalFields: Field[]
}

export type CameraOptions = {
  center: [number, number],
  zoom: number,
  bearing: number,
  pitch: number
}

export type CommonViewContentProps = {
  /** Array of observations to render */
  observations: Array<Observation>,
  /** Called with id of observation clicked, optionally with image index */
  onClick: (id: string, imageIndex?: number) => void,
  getPreset: Observation => PresetWithAdditionalFields,
  /**
   * For a given attachment, return `src` and `type`
   */
  getMedia: GetMedia
}
